pyquil/quilbase.py

```diff
@@ -218,13 +218,11 @@ def _convert_to_py_instruction(instr: Any) -> AbstractInstruction:
     if isinstance(instr, quil_rs.Declaration):
         return Declare._from_rs_declaration(instr)
     if isinstance(instr, quil_rs.Delay):
-        print("len(instr.qubits):", len(instr.qubits))
-        print("len(instr.frame_names):", len(instr.frame_names))
-        if len(instr.qubits) > 0 and len(instr.frame_names) > 0:
-            return Delay._from_rs_delay(instr)
+        if len(instr.frame_names) > 0:
+            return DelayFrames._from_rs_delay(instr)
         if len(instr.qubits) > 0:
             return DelayQubits._from_rs_delay(instr)
-        return DelayFrames._from_rs_delay(instr)
+        return Delay._from_rs_delay(instr)
```

Comment:
`quil-rs` only defines a single [Delay](https://github.com/rigetti/quil-rs/blob/main/quil-rs/src/instruction/timing.rs#L5-L9) type that can include both Qubits and frames. And really, [per the Quil spec](https://quil-lang.github.io/#12-7Timing-and-Synchronization) a `DELAY` should always have at least 1 qubit. This means, as I wrote it previously, `DelayFrames` should never be returned. With this change a `Delay` would only be returned if it had no frames _or_ qubits, which isn't really valid. 

I think pyQuil is modeling the spec wrong in this case, and we're kind of stuck doing our best here. In a v5 release, I think it would make sense to get rid of `DelayFrame`s and `DelayQubit`s entirely.

Reply:
That's a good idea, referring to the Quil Spec (which I did not think to look at on Friday).

The Quil Spec says a `DELAY` has 1 or more qubits and zero or more frame identifiers.

PyQuil v3 models this by having 
1. `DelayFrames` for the case where there are frame identifiers and 
2. `DelayQubits` for the case where there are not and 
3. no generic `Delay` type because there is no third case.
4. The function `DELAY` returns either #1 or #2 depending on whether you pass it a list of frames or a list of qubits.

If I understand correctly, there is nothing preventing us from following exactly the same rules in PyQuil v4:
- The function `DELAY` already behaves the same way, I think.
- The compatibility layer can return `DelayFrames` if there are frame identifiers and `DelayQubits` otherwise, and never return `Delay` which doesn't exist in v3.

If we choose not to maintain this level of compatibility then
1. I think it should be documented along with the other breaking changes and
2. Whay not just jump straight to the forecasted v5, single-delay-type world?

Reply:
Yeah, upon reflection, I think you are right. Let's go ahead and move forward with this change if it fixes your issue downstream. I'll consider deprecation of DelayFrames/Qubits as a follow-up. I don't want to jump to a v5 release quite yet.